### PR TITLE
koji_upload: add to render in build_request

### PR DESCRIPTION
**Pull request by mlangsdorf**
add render_koji_upload to build_request.

Signed-off-by: Mark Langsdorf <mlangsdo@redhat.com>

### CODE REVIEW

File: osbs/build/build_request.py [hunk 1]
```diff
@@ -680,6 +680,31 @@ def render_koji_promote(self, use_auth=None):
                         "specified")
             self.dj.remove_plugin("exit_plugins", "koji_promote")
 
+# FIXME edit parameters and stuff
+    def render_koji_upload(self, use_auth=None):
+        if not self.dj.dock_json_has_plugin_conf('postbuild_plugins',
+                                                 'koji_upload'):
+            return
+
+        if self.spec.kojihub.value:
```
Comment (lcarva) [inline osbs/build/build_request.py]: Also need to set `build_json_dir`. See how `render_import_image` does it.

File: inputs/prod_inner.json [hunk 2]
```diff
@@ -124,6 +124,17 @@
...
+      "args": {
+        "kojihub": "{{KOJI_HUB}}",
+        "url": "{{OPENSHIFT_URI}}",
+        "verify_ssl": false,
+        "blocksize": 10485760,
+        "koji_keytab": false,
+        "koji_principal": false
+      }
+    },
+    {
```
Comment (twaugh) [inline inputs/prod_inner.json]: This needs to go in the worker inner JSON for the new arrangement, not in prod_inner.json, and it also needs to run after the "compress" plugin. Best is between "compress" and "pulp_pull" I think.

File: osbs/build/build_request.py (hunk 1 above)
Comment (mlangsdorf) [inline osbs/build/build_request.py]: Fixed.

File: inputs/prod_inner.json (hunk 2 above)
Comment (mlangsdorf) [inline inputs/prod_inner.json]: Fixed I think.

**Comment by twaugh** [comment]
We'll also want to make sure to remove the koji upload/import plugins for scratch builds (in `adjust_for_scratch`).

File: inputs/prod_inner.json (hunk 2 above)
Comment (twaugh) [inline inputs/prod_inner.json]: This still modifies prod_inner.json.

**Comment by twaugh** [comment]
Be aware that https://github.com/projectatomic/osbs-client/pull/623 adds unit testing for arrangement versions. Once it's merged, it would be worth re-basing this PR and extending those tests.

File: osbs/build/build_request.py [hunk 3]
```diff
@@ -1071,6 +1091,12 @@ def render(self, validate=True):
                           ('exit_plugins', 'koji_promote', 'koji_ssl_certs'):
                           self.spec.koji_certs_secret.value,
 
+                          ('postbuild_plugins', 'koji_upload', 'koji_ssl_certs_dir'):
+                          self.spec.koji_certs_secret.value,
+
+                          ('exit_plugins', 'koji_import', 'koji_ssl_certs'):
```
Comment (lcarva) [inline osbs/build/build_request.py]: Is doing this for `koji_import` premature? Development of this plugin has not been started, correct?

File: osbs/build/build_request.py [hunk 4]
```diff
@@ -681,6 +682,25 @@ def render_koji_promote(self, use_auth=None):
...
+
+        self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload', 'url',
+                                  self.spec.builder_openshift_url.value)
+        if self.spec.kojihub.value:
+            self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload',
+                                      'kojihub', self.spec.kojihub.value)
+            if use_auth is not None:
+                self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload',
+                                          'use_auth', use_auth)
+
```
Comment (lcarva) [inline osbs/build/build_request.py]: Also needs to set `build_json_dir` param. See `render_import_image` method.

File: tests/build_/test_build_request.py [hunk 5]
```diff
@@ -151,6 +151,19 @@ def assert_import_image_plugin(self, plugins, name_label, registry_uri,
                                  "postbuild_plugins", "import_image", "args",
                                  "insecure_registry")
 
+    def assert_koji_upload_plugin(self, plugins, use_auth):
```
Comment (twaugh) [inline tests/build_/test_build_request.py]: This method is not used.

**Comment by twaugh** [comment]
There are a couple of added lines of code that don't get coverage testing:
https://coveralls.io/builds/12258027/source?filename=osbs%2Fbuild%2Fbuild_request.py#L703

Other than that, this looks really good.

File: osbs/build/build_request.py [hunk 6]
```diff
@@ -681,6 +682,28 @@ def render_koji_promote(self, use_auth=None):
...
+        self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload', 'url',
+                                  self.spec.builder_openshift_url.value)
+        if self.spec.kojihub.value:
+            self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload',
+                                      'kojihub', self.spec.kojihub.value)
+            if use_auth is not None:
+                self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload',
+                                          'use_auth', use_auth)
+                self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload',
```
Comment (lcarva) [inline osbs/build/build_request.py]: `build_json_dir` is independent of `use_auth` value.

File: osbs/build/build_request.py [hunk 7]
```diff
@@ -681,6 +682,24 @@ def render_koji_promote(self, use_auth=None):
                         "specified")
             self.dj.remove_plugin("exit_plugins", "koji_promote")
 
+    def render_koji_upload(self, use_auth=None):
+        if not self.dj.dock_json_has_plugin_conf('postbuild_plugins', 'koji_upload'):
+            return
+
+        self.dj.dock_json_set_arg('postbuild_plugins', 'koji_upload', 'url',
+                                  self.spec.builder_openshift_url.value)
```
Comment (twaugh) [inline osbs/build/build_request.py]: Odd not to put this inside the blocked below, but I guess it's fine.

Comment (lcarva) [inline osbs/build/build_request.py]: It's now being set twice?

Comment (mlangsdorf) [inline osbs/build/build_request.py]: dangit.

**Comment by twaugh** [comment]
Draft release notes updated.

**Pull request merged**